src/bin/pg_autoctl/supervisor.c

```diff
@@ -34,12 +34,28 @@
 #include "signals.h"
 #include "string_utils.h"
 
+/*
+ * A static struct to loop seamlessly through the active services in a
+ * supervisor struct. Those can be either 'static' or 'dynamic'.
+ */
```

Comment:
Terribly sorry but I am failing to see the point here. This is just a struct.
Are you objecting to having an iterator or how the iterator is implemented?

Reply:
What I mean is that you then have introduced the following code when using the macro. Where we used to return a pointer to an existing data structure, we now return a copy of it, done in two steps, one with a static allocation on the stack in the iterator internal function and second step with a call to `memcpy`. Why is that a preferred way of operating here? Who is going to be responsible for freeing that new allocated memory?

``` diff
-			*result = &(supervisor->services[serviceIndex]);
+ 			*result = service;
```

My feeling is that we should continue as before, where this memory handling is obvious, because that's where the bugs in C are coming from first. Having to write another same loop again is rarely the source of bugs in production.

I yeah I am objecting to having an iterator abstraction altogether here.

Reply:
I deeply apologise. I really do not mean to be contradictory. I am simply a bit at a loss here.

Where did you see this memcpy? Is it on this diff? The iterator does not do any memcpy() nor the macro.

These are the iterator functions:

```
+static inline Service *
+supervisor_it_get(Supervisor_it *sit)
+{
+       if (sit->index < sit->count)
+       {
+               return &sit->services[sit->index];
+       }
+
+       return NULL;
+}
+
+
+static inline Service *
+supervisor_it_first(Supervisor_it *sit, Supervisor *supervisor)
+{
+       *sit = (Supervisor_it) {
+               .supervisor = supervisor,
+               .services = supervisor->services,
+               .count = supervisor->serviceCount,
+               .index = 0,
+       };
+
+       return supervisor_it_get(sit);
+}
+
+
+static inline Service *
+supervisor_it_next(Supervisor_it *sit)
+{
+       sit->index++;
+
+       /* if we are done with the static services, move to the dynamic */
+       if (sit->index >= sit->count &&
+               sit->services == sit->supervisor->services)
+       {
+               sit->services = sit->supervisor->dynamicServices.services;
+               sit->count = sit->supervisor->dynamicServices.serviceCount;
+               sit->index = 0;
+       }
+
+       return supervisor_it_get(sit);
+}
``` 

And this is the macro:
```
+#define foreach_supervised_service(it, supervisor, service) \
+       for ((service) = supervisor_it_first((it), (supervisor)); \
+                (service) != NULL; \
+                (service) = supervisor_it_next(it))
```

I will happily remove the macro, no problem. I simply fail to see the memcpy here.

Again terribly sorry for this and thank you very much for your time.

Reply:
Ah yes, indeed, the `memcpy` are used in parts of the code that have nothing to do with the iterators. I still find the iterator very confusing. It's a lot of code to replace 3 loops that are very easy to read, so I keep failing to see the point. It's one more thing to keep in mind and juggle with, it doesn't make anything simpler, just 3 call sites only a little easier to write. Let's remove the iterator idea, even though that's not having the problems I though it had, because I think it's just too much for the list of changes in this PR.

Reply:
Now I understand. Thank you for your patience and time.

I will prepare something to show you how it would like without
a) the macro
b) the macro and the iterator
and let you choose what you prefer.

Reply:
Sure, but let's leave this last. Once everything else is agreed upon I can give you the two options to choose from.

I am really not married to the idea. No worries